src/runtime-rs/crates/hypervisor/src/ch/inner_hypervisor.rs

```diff
@@ -761,17 +762,116 @@ impl CloudHypervisorInner {
+        self.guest_memory_block_size_mb = size;
     }
 
     pub(crate) fn guest_memory_block_size_mb(&self) -> u32 {
-        self._guest_memory_block_size_mb
+        self.guest_memory_block_size_mb
     }
 
-    pub(crate) fn resize_memory(&self, _new_mem_mb: u32) -> Result<(u32, MemoryConfig)> {
-        warn!(sl!(), "CH memory resize not implemented - see https://github.com/kata-containers/kata-containers/issues/8801");
+    pub(crate) async fn resize_memory(
+        &mut self,
+        mut new_mem_mb: u32,
+    ) -> Result<(u32, MemoryConfig)> {
+        // check for probe in memory_config
+        let caps = &self.capabilities;
+        if caps.is_mem_hotplug_probe_supported() {
+            warn!(sl!(), "probe memory is not supported for cloud hypervisor");
+            return Ok((0, MemoryConfig::default()));
+        }
+
+        // check for resizing to 0
+        if new_mem_mb == 0 {
+            error!(sl!(), "Cannot resize memory to 0");
+            return Ok((0, MemoryConfig::default()));
+        }
+
+        // get hotplug size
+        let max_hotplug_size_mb = self.hypervisor_config().memory_info.default_maxmemory;
+        // do we want to set new_mem_mb to hotplug_size or fail?
+        if new_mem_mb > max_hotplug_size_mb {
+            warn!(sl!(), "Memory size unchanged; requested memory size {} is greater than max memory size {}", new_mem_mb, max_hotplug_size_mb);
+            return Ok((0, MemoryConfig::default()));
+        }
+
+        // get mem size
+        let default_mem_size_mb = self.hypervisor_config().memory_info.default_memory;
+        let current_hotplug_size_mb = self.mem_hotplug_size_mb;
+        let current_mem_size_mb = default_mem_size_mb + current_hotplug_size_mb;
+
+        if new_mem_mb == current_mem_size_mb {
+            info!(sl!(), "VM already has requested memory size {}", new_mem_mb);
+            return Ok((current_mem_size_mb, MemoryConfig::default()));
+        }
+
+        if new_mem_mb < current_mem_size_mb {
+            warn!(sl!(), "Remove memory is not supported, nothing to do");
+            return Ok((current_mem_size_mb, MemoryConfig::default()));
+        }
+
+        let block_size_mb = self.guest_memory_block_size_mb;
+
+        let new_hotplug_size_mb = new_mem_mb - current_mem_size_mb;
+        let aligned_new_hotplug_size_mb =
+            checked_next_multiple_of(new_hotplug_size_mb.into(), block_size_mb.into());
+
+        if let Some(aligned_value_mb) = aligned_new_hotplug_size_mb {
+            let aligned_request = current_mem_size_mb + aligned_value_mb as u32;
+            if new_mem_mb != aligned_request {
+                info!(
+                    sl!(),
+                    "Aligning VM memory request {} to {} with block size {}", new_mem_mb, aligned_request, block_size_mb;
+                );
+                new_mem_mb = aligned_request;
+            }
+        }
+
+        // Additional checks after memory alignment
```

Comment:
Hi @cmaf 

Is the init default memory  block size aligned? If the initial values are aligned to the block size, then it is enough to do the alignment first and then do these size checks. Is there no need to check before the alignment?

Reply:
That's a very interesting question.  Correct me if I'm wrong but it would be hard to rely on `memory_info.default_memory` being aligned since 1) it comes from the config file, ie. the user, and 2) the memory allocation granularity (the block size) is only determined at runtime via the agent.

So I _think_ (and I might well be off here) that at least on the initial invocation of this function when `current_hotplug_size_mb` is likely zero and `current_mem_size_mb` thus consists purely of `memory_info.default_memory` alignment of the initial values can hardly be guaranteed.

That said, I still think you're right suggesting that the pre-alignment checks could be left out.  They seem to serve as an early-out to avoid the alignment computations if it's clear that they are not needed.  This would make a lot of sense if they were considerably resource-heavy.  However, since that doesn't seem the case I'd consider leaving the pre-alignment checks out, simplifying the function at the risk that a bunch of simple arithmetics might be performed unnecessarily at times.

Reply:
@lifupan @pmores I switched the checks to only occur after alignment